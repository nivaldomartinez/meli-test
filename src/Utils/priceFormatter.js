/**
 * number to currency format
 * @param {*} price 
 * @returns 
 */
const priceFormatter = (price) => {
    const formattedPrice = new Intl.NumberFormat().format(price)
    return formattedPrice
}

export default priceFormatter;