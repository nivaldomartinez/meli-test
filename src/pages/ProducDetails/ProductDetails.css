.product-details {
    background-color: white;
    border-radius: 5px;
}

.details-container {
    display: grid;
    padding: 32px;
    grid-template-columns: 680px 1fr;
    gap: 16px;
}

.product-detail-info {
    color: #333;
}

.product-detail-condition {
    font-size: 14px;
    color: #666;
    margin-bottom: 16px;
}

.product-detail-image {
    width: 680px;
    height: 500px;
    border-radius: 4px;
    background-position: center;
    background-size: contain;
    background-repeat: no-repeat;
}

.product-detail-price {
    font-size: 46px;
    display: flex;
    align-items: flex-start;
    gap: 5px;
    padding-top: 32px;
    padding-bottom: 32px;
}

.product-detail-title {
    font-size: 24px;
}

.amount-decimals {
    font-size: 20px;
}

.btn-buy {
    background-color: #4383FA;
    padding: 10px;
    display: flex;
    justify-content: center;
    border-radius: 5px;
    color: white;
    font-weight: 600;
    cursor: pointer;
}

.btn-buy:hover {
    background-color: #528dfa;
}

.product-detail-description {
    width: calc(680px - 36px);
    padding: 36px;
    text-align: justify;
    text-justify: inter-word;
}

.title-description {
    font-size: 28px;
}

.description {
    font-size: 16px;
    color: #999;
}