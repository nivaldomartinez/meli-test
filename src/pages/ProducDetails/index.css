.product-details {
    background-color: white;
    border-radius: 5px;
}

.details-container {
    display: flex;
    justify-content: space-between;
    padding: 32px;
}

.product-detail-info {
    color: #333;
}

.product-detail-condition {
    font-size: 14px;
    color: #666;
    margin-bottom: 16px;
}

.product-detail-image {
    width: 680px;
    height: 500px;
    border-radius: 4px;
    background-position: center;
    background-size: cover;
    background-repeat: no-repeat;
}

.product-detail-price {
    font-size: 46px;
    display: flex;
    align-items: flex-start;
    gap: 5px;
    padding-top: 32px;
    padding-bottom: 32px;
}

.product-detail-title {
    font-size: 24px;
}

.amount-decimals {
    font-size: 20px;
}

.btn-buy {
    background-color: #4383FA;
    padding: 10px;
    display: flex;
    justify-content: center;
    border-radius: 5px;
    color: white;
    font-weight: 600;
}

.product-detail-description {
    padding: 36px;
}

.title-description {
    font-size: 28px;
}

.description {
    font-size: 16px;
    color: #999;
}