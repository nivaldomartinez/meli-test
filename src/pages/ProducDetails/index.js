import React from 'react';
import { useParams } from 'react-router';
import './index.css';

const ProductDetails = () => {
    const { productId } = useParams()

    const product = {
        id: '8721934782342',
        title: 'Xiaomi redmi note 9',
        price: {
            currency: '$',
            amount: 680000,
            decimals: '00'
        },
        picture: 'https://i.blogs.es/3a3421/xiaomi-redmi-note-10-0/1366_2000.jpeg',
        condition: 'Nuevo',
        free_shipping: false,
        description: 'Description del user',
        sold_quantity: 234
    }

    return (
        <div class="product-details">
            <div className="details-container">
                <div className="product-detail-image" style={{ backgroundImage: `url(${product.picture})` }}></div>
                <div className="product-detail-info">
                    <div className="product-detail-condition">
                        <span>
                            {product.condition} - {product.sold_quantity} vendidos
                        </span>
                    </div>
                    <div className="product-detail-title">
                        <span>{product.title}</span>
                    </div>
                    <div className="product-detail-price">
                        <span>{product.price.currency} {product.price.amount}</span> <span className="amount-decimals">{product.price.decimals}</span>
                    </div>
                    <div className="btn-buy">
                        <span>Comprar</span>
                    </div>
                </div>
            </div>
            <div className="product-detail-description">
                <span className="title-description">Descripcion del producto</span>
                <p className="description">{product.description}</p>
            </div>
        </div>
    )
}

export default ProductDetails;