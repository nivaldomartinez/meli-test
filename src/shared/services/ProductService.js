import axios from "axios";
import endpoints, { BASE_URL } from "../../Utils/endpoints";
import replaceParams from "../../Utils/replaceParams";

const intance = axios.create({
    baseURL: BASE_URL
})

/**
     * get product details from backend
     */
export const getProductDetails = (productId) => {
    return intance.get(replaceParams(endpoints.productDetail, [productId]))
}

/**
 * get breadcrumb data from server
 * @param {*} categoryId product categoryId
 */
export const getBreadcrumb = (categoryId) => {
    return intance.get(replaceParams(endpoints.breadcrumb, [categoryId]))
}

/**
* get products from api
* @param {*} location router location to check changes from the url param
*/
export const getProducts = (query) => {
    return intance.get(replaceParams(endpoints.search, [query]));
}