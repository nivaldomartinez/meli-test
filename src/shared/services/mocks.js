export const searchResponseMock = {
    "author": {
        "name": "MELI",
        "lastname": "TEST"
    },
    "categories": [
        "MLA3422",
        "MLA3422",
        "MLA3422",
        "MLA3422"
    ],
    "items": [
        {
            "id": "MLA903952589",
            "title": "Muñeco Figura Soft Toy Story Woody Buzz Jessie 18cm Disney",
            "price": {
                "currency": "ARS",
                "amount": 1563,
                "decimals": 15
            },
            "picture": "http://http2.mlstatic.com/D_817618-MLA46821124189_072021-O.jpg",
            "condition": "new",
            "free_shipping": false
        },
        {
            "id": "MLA668912898",
            "title": "Muñeco Buzz Lightyear Toy Story Soft Original Disney ",
            "price": {
                "currency": "ARS",
                "amount": 2285,
                "decimals": 0
            },
            "picture": "http://http2.mlstatic.com/D_602827-MLA46232786132_062021-I.jpg",
            "condition": "new",
            "free_shipping": false
        },
        {
            "id": "MLA886350340",
            "title": "Juguete Muñeco Figura Articulada Toy Story 4 Personajes",
            "price": {
                "currency": "ARS",
                "amount": 790,
                "decimals": 56
            },
            "picture": "http://http2.mlstatic.com/D_798961-MLA46942035808_082021-O.jpg",
            "condition": "new",
            "free_shipping": false
        },
        {
            "id": "MLA606398749",
            "title": "Woody Buzz Jessie New Toys 40cm Alto Con Sombrero",
            "price": {
                "currency": "ARS",
                "amount": 2499,
                "decimals": 0
            },
            "picture": "http://http2.mlstatic.com/D_600276-MLA46855684106_072021-I.jpg",
            "condition": "new",
            "free_shipping": false
        }
    ]
}

export const breadcrumbResponseMock = {
    "author": {
        "name": "MELI",
        "lastname": "TEST"
    },
    "breadcrumb": [
        {
            "id": "MLA1132",
            "name": "Juegos y Juguetes"
        },
        {
            "id": "MLA432888",
            "name": "Muñecos y Muñecas"
        },
        {
            "id": "MLA3422",
            "name": "Muñecos y Figuras de Acción"
        }
    ]
}

export const productDetailMock = {
    "author": {
        "name": "MELI",
        "lastname": "TEST"
    },
    "item": {
        "id": "MLA730906646",
        "title": "Vinilo Autoadhesivo Impreso Toy Story Catalogo De Imagenes",
        "price": {
            "currency": "ARS",
            "amount": 330,
            "decimals": 75
        },
        "picture": "http://http2.mlstatic.com/D_828728-MLA27545108086_062018-O.jpg",
        "condition": "new",
        "free_shipping": false,
        "sold_quantity": 5,
        "description": "vinilo autoadhesivo impreso toy story catalogo de imagenes para impresión de murales pared y puertas precio por 0.50 X 0.50\nconsulta tus medidas especiales",
        "category_id": "MLA117185"
    }
}