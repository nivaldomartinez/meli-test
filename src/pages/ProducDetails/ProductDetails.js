import React, { Fragment, useEffect, useState } from 'react';
import { Helmet } from 'react-helmet';
import { useParams } from 'react-router';
import { getBreadcrumb, getProductDetails } from '../../shared/services/ProductService';
import priceFormatter from '../../Utils/priceFormatter';
import './ProductDetails.css';

const ProductDetails = () => {
    const { productId } = useParams()

    const [product, setProduct] = useState();
    const [breadcrumb, setBreadcrumb] = useState();

    useEffect(() => {
        getDetails();
    }, [])

    /**
     * get product details from backend
     */
    const getDetails = () => {
        getProductDetails(productId).then(productResponse => {
            const { item: product } = productResponse.data;
            setProduct(product);
            getBreadcrumbData(product.category_id)
        }).catch(error => { });
    }

    /**
     * get breadcrumb data by category from server
     * @param {*} categoryId 
     */
    const getBreadcrumbData = (categoryId) => {
        getBreadcrumb(categoryId).then(breadcrumbResponse => {
            setBreadcrumb(breadcrumbResponse.data.breadcrumb)
        }).catch(error => { });
    }

    return (
        product ?
            <Fragment>
                <Helmet>
                    <title>{product.title} | Meli Test</title>
                </Helmet>
                <div className="breadcrumb" id="details-breadcrumb">
                    {breadcrumb ? breadcrumb.map(item => {
                        return (
                            <div className="item" key={item.id}>
                                <span>{item.name}</span>
                            </div>
                        )
                    }) : null}
                </div>
                <div className="product-details">
                    <div className="details-container">
                        <div className="product-detail-image" id="picture"
                            style={{ backgroundImage: `url(${product.picture})` }}></div>
                        <div className="product-detail-info">
                            <div className="product-detail-condition">
                                <span>
                                    {product.condition} - {product.sold_quantity} vendidos
                                </span>
                            </div>
                            <div className="product-detail-title">
                                <span>{product.title}</span>
                            </div>
                            <div className="product-detail-price">
                                <span>${priceFormatter(product.price.amount)}</span> <span className="amount-decimals">{product.price.decimals}</span>
                            </div>
                            <div className="btn-buy">
                                <span>Comprar</span>
                            </div>
                        </div>
                    </div>
                    <div className="product-detail-description">
                        <span className="title-description">Descripcion del producto</span>
                        <p className="description">{product.description}</p>
                    </div>
                </div>
            </Fragment> : null
    )
}

export default ProductDetails;