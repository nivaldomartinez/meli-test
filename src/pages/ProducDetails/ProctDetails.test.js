import '@testing-library/jest-dom/extend-expect';
import { cleanup, render, waitFor } from '@testing-library/react';
import { createMemoryHistory } from "history";
import React, { useState } from 'react';
import { Route, Router } from 'react-router-dom';
import * as ProductService from '../../shared/services/ProductService';
import priceFormatter from '../../Utils/priceFormatter';
import ProducDetails from './ProductDetails';
jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useState: jest.fn()
}))

const productMock = {
    id: "MLA633913166",
    title: "Toy Story Buzz Lightyear Peluche Original",
    price: {
        currency: "ARS",
        amount: 3000,
        decimals: 15
    },
    picture: "http://http2.mlstatic.com/D_999725-MLA42581294630_072020-O.jpg",
    condition: "new",
    free_shipping: false,
    sold_quantity: 25,
    description: "PELUCHE SOFT BUZZ LIGHTYEAR\nOriginal New Toys!\n\n\n40cm aproximadamente.",
    category_id: "MLA1166"
}

const breadcrumbMock = [
    { id: "MLA1132", name: "Juegos y Juguetes" },
    { id: "MLA432888", name: "Mu침ecos y Mu침ecas" },
    { id: "MLA3422", name: "Mu침ecos y Figuras de Acci칩n" }
]


const renderWithRouter = (component) => {
    const history = createMemoryHistory({
        initialEntries: ["/items/MLA606398749"],
    });
    const Wrapper = ({ children }) => (
        <Router history={history}>
            <Route path="/items/:productId">{children}</Route>
        </Router>
    );
    return {
        ...render(component, { wrapper: Wrapper }),
        history,
    };
};


describe('testing ProductDetails.js', () => {

    beforeEach(() => {
        useState.mockImplementation(jest.requireActual('react').useState);
    })

    afterAll(cleanup)

    test('<ProductDetails/> should be rendered', () => {

        const component = renderWithRouter(<ProducDetails />)
        expect(component.container).toBeEmptyDOMElement();
        expect(component).toBeDefined();
    })

    test('<ProductDetails/> should render breadcrumb and match snapshot', () => {

        useState
            .mockImplementationOnce(() => [productMock, jest.fn()])
            .mockImplementationOnce(() => [breadcrumbMock, jest.fn()])

        const component = renderWithRouter(<ProducDetails />)
        component.getAllByText(breadcrumbMock[0].name);

        expect(component.container).toMatchSnapshot();

    })

    test('<ProductDetails/> should render product properties', async () => {

        jest.spyOn(ProductService, 'getProductDetails').mockResolvedValue({ data: { item: productMock } });
        jest.spyOn(ProductService, 'getBreadcrumb').mockResolvedValue({ data: { breadcrumb: breadcrumbMock } });

        const component = renderWithRouter(<ProducDetails />)
        await waitFor(() => component.getAllByText(productMock.title));
        await waitFor(() => component.getAllByText(productMock.title));
        await waitFor(() => component.getAllByText(productMock.price.decimals))
        await waitFor(() => component.getAllByText(`$${priceFormatter(productMock.price.amount)}`))
        await waitFor(() => component.getAllByText(`${productMock.condition} - ${productMock.sold_quantity} vendidos`))
        await waitFor(() => expect(component.container.querySelector('#picture')).toHaveStyle({
            'background-image': `url(${productMock.picture})`
        }));

    })

    test('<ProductDetails/> should render empty if getProductDetails fails', async () => {

        jest.spyOn(ProductService, 'getProductDetails').mockRejectedValue({});

        const component = renderWithRouter(<ProducDetails />)
        expect(component.container).toBeEmptyDOMElement();

    })

    test('<ProductDetails/> should render empty breadcrumb if getBreadcrumb fails', async () => {

        jest.spyOn(ProductService, 'getProductDetails').mockResolvedValue({ data: { item: productMock } });
        jest.spyOn(ProductService, 'getBreadcrumb').mockRejectedValue({});

        const component = renderWithRouter(<ProducDetails />)
        await waitFor(() => component.getAllByText(productMock.title));
        expect(component.container.querySelector('#details-breadcrumb')).toBeEmptyDOMElement();

    })
})