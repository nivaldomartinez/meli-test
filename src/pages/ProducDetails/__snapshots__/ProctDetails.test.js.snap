// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing ProductDetails.js <ProductDetails/> should render breadcrumb and match snapshot 1`] = `
<div>
  <div
    class="breadcrumb"
    id="details-breadcrumb"
  >
    <div
      class="item"
    >
      <span>
        Juegos y Juguetes
      </span>
    </div>
    <div
      class="item"
    >
      <span>
        Mu침ecos y Mu침ecas
      </span>
    </div>
    <div
      class="item"
    >
      <span>
        Mu침ecos y Figuras de Acci칩n
      </span>
    </div>
  </div>
  <div
    class="product-details"
  >
    <div
      class="details-container"
    >
      <div
        class="product-detail-image"
        id="picture"
        style="background-image: url(http://http2.mlstatic.com/D_999725-MLA42581294630_072020-O.jpg);"
      />
      <div
        class="product-detail-info"
      >
        <div
          class="product-detail-condition"
        >
          <span>
            new
             - 
            25
             vendidos
          </span>
        </div>
        <div
          class="product-detail-title"
        >
          <span>
            Toy Story Buzz Lightyear Peluche Original
          </span>
        </div>
        <div
          class="product-detail-price"
        >
          <span>
            $
            3,000
          </span>
           
          <span
            class="amount-decimals"
          >
            15
          </span>
        </div>
        <div
          class="btn-buy"
        >
          <span>
            Comprar
          </span>
        </div>
      </div>
    </div>
    <div
      class="product-detail-description"
    >
      <span
        class="title-description"
      >
        Descripcion del producto
      </span>
      <p
        class="description"
      >
        PELUCHE SOFT BUZZ LIGHTYEAR
Original New Toys!


40cm aproximadamente.
      </p>
    </div>
  </div>
</div>
`;
